### PR TITLE
Tokio migration

### LINKED ISSUE #368: Updating url dependency
Hey everyone!

Is there a particular reason why you guys stick to the 1.7.x versions of the `url` crate? I am using `librespot` in a project where I need to use the new 2.x version as the project uses serde 1.x.

If not, I would gladly create a PR to update the dependencies! :)

### LINKED ISSUE #596: Move get_credentials function to the application
I would like to suggest another breaking change before 0.2: removing `fn get_credentials`. Why? It does not provide any functionality directly concerning Spotify, but it's rather a very specialized function, written with a concrete application/daemon in mind.

Yes, spotifyd uses this function too, or better: they abuse it. Instead of prompting in the `prompt` callback they abort the application. Which shows that this function is too specialized for their usecase too. 

So I would suggest moving it to the application. It's not hard to implement, so every dependency using this function could implement it's own more spezialized function. Do you agree?

### LINKED ISSUE #677: Memory Leak?
When running Librespot (release mode) and skipping through playlists, the used memory increases slowly but steady. It starts with 2-3MB, but after a while I was able to reach 8MB. 

It seems to happen in both the `dev` and the `tokio_migration` branch.

### PR DESCRIPTION
**Pull request by Johannesd3**
We're very close to finishing the migration to Tokio 1.0, and I think the result is worth seeing. Not only tokio but almost every other dependency was bumped to its latest version. In many places the code became simpler by using the async/await notation, and there were some general efforts of refactoring the code on this occasion.

__Issues:__
- [x] The examples do not compile (#656)
- [x] Proxy support is still a bit unclear (#653)

__Pull requests:__
- [x] clean up imports & clippy lints (#652)
- [x] Refactor AudioFileFetch (#658)
- [x] Update examples (#663)
- [x] Fix remaining clippy warnings (#667)
- [x] Add back hyper-proxy (#674)
- [x] Further progress on tokio migration (#687) 

__Things that were removed and might or might not be added back:__
- [x] `Default` impl of `SessionConfig` (was added again in #663)
- [ ] `LinearMap` dependency in `librespot-metadata` (replaced by `HashMap` in a6ed685)
- [ ] `fn get_credentials` (removed in 9253be7, see #596)

__and__
- [ ] ~Think about feature flags (see #648, #652)~ (restored old behaviour and postponed)
- [x] Solve merge conflicts (#683)

Fixes #677, fixes #596, fixes #368

### CODE REVIEW

**Review by Johannesd3 — COMMENTED** [review]
I've written down some things I noticed as I went through the changes.

**Comment by raftario** [comment]
Anything I could do to help with this ?

**Comment by Johannesd3** [comment]
I brought up a few points where I'm not sure whether I made the right decision, but after #674 I would consider this branch as ready to be merged into dev.

**Comment by Johannesd3** [comment]
@ashthespy Would you try to merge dev in?

**Comment by ashthespy** [comment]
@Johannesd3 I'm away from a console for a bit longer, sorry!

File: playback/src/player.rs [hunk 1]
```diff
@@ -317,22 +318,21 @@ impl Player {
...
-            .get_player_event_channel()
-            .filter(|event| match event {
-                PlayerEvent::EndOfTrack { .. } | PlayerEvent::Stopped { .. } => true,
-                _ => false,
-            })
-            .into_future()
-            .map_err(|_| ())
-            .map(|_| ());
-        Box::new(result)
+    pub async fn await_end_of_track(&self) {
```
[the remark this answers is not in the document]
Comment (k3d3) [reply playback/src/player.rs:325]: I'd probably name it to end_of_track - I feel like calling `await_end_of_track().await` would be a bit weird.

[the remark this answers is not in the document]
Comment (Johannesd3) [reply playback/src/player.rs:325]: I totally get your point, but if you'd see a function `end_of_track` in the docs, would you know what it does?

**Comment by sashahilton00** [comment]
FWIW I've been running this branch without issue for about a week now. Given that the remaining todos are mostly think about/do we need items, and since #674 is now merged, now is probably as good a time as any to merge this into dev and release `0.2.0`. Unless there are any objections, once this is rebased I'll merge it.

**Comment by Johannesd3** [comment]
I certainly won't rebase 74 commits, we'll have to merge the changes of dev in. #687 contains dev until yesterday, but the latest PR by @roderickvd will be hard to merge ...

File: connect/src/spirc.rs [hunk 2]
```diff
@@ -317,143 +320,107 @@ impl Spirc {
...
                     }
-                }
-            }
-
-            let poll_sender = self.sender.poll_complete().unwrap();
-
-            // Only shutdown once we've flushed out all our messages
-            if self.shutdown && poll_sender.is_ready() {
-                return Ok(Async::Ready(()));
+                },
```
[the remark this answers is not in the document]
Comment (Johannesd3) [reply connect/src/spirc.rs:414]: @sashahilton00 did you also make use of those feature while testing it? I don't really know what they're for.

**Comment by sashahilton00** [comment]
Well this PR is a monster, quite the achievement. Thanks for putting the work in. Are we happy to merge?

**Comment by Johannesd3** [comment]
There's no reason for it, but I'm still anxious. Maybe just wait 3 days, so everyone might test their own favourite features again after the lastest changes, but there are no outstanding or planned changes from my side.

**Pull request merged**